### PR TITLE
check_disk: add ignore-missing option to return OK for missing fs

### PR DESCRIPTION
**Pull request by KriSchu**
There a situations where UNKNOWN or CRITICAL services are not wanted when a filesystem is missing, a regex does not match or the filesystem is inaccessible on a system.
This new option helps to have the service in state OK.

Potentially solves 
#842 
#847 
#1494 (maybe this issue as well)

### CODE REVIEW

**Comment by sni** [comment]
that's a nice addition. Could you by chance add a test case for this new option?

**Comment by KriSchu** [comment]
@sni I tried, but I can't get the environment prepared to run the tests. Unfortunately there is no howto in the README.
So I'm not able to check if my tests work and I don't want to push untested code.

**Comment by KriSchu** [comment]
@sni I added tests yesterday and hope they work. Can you trigger the checks/ tests?

**Comment by sni** [comment]
great. Looks good. You only need to adjust the test plan.

**Comment by KriSchu** [comment]
@sni I don't know what to do here. 
It took me way more time already to try to figure out how these tests work or where the test plans are described, than I anticipated (expected to quickly help out with a fix).

Can you solve the tests maybe? I'll have a look at your changes and try to learn from that. I hit the checkbox that Maintainers are able to make changes.

Thanks in advance.

**Comment by sni** [comment]
no worries, i just updated the file. Lets see if that's it already.

**Review by sni — APPROVED** [review]

**Comment by sni** [comment]
Thanks for this feature and the test case.

**Comment by KriSchu** [comment]
Can you let me know how the process is regarding merges and new releases? When will this PR be merged and included in a future release approximately?

**Comment by RincewindsHat** [comment]
I don't want to be annoying, but I tried this a little bit and I was a little bit irritated.
The plugin still quits when a file system is missing (but says it is ok). So lets say I would like to do something like this: `check_disk -w 10% -c 5% -p / -p /home -p /foo` (where `/foo` does not exist), I would expect the plugin to perform the plugin to perform the tests on `/` and `/home` normally (and modify the result accordingly) and not quit and throw the towel.

**Comment by KriSchu** [comment]
Hi @RincewindsHat 
Good catch! Thanks a lot for that! We totally missed that in the used test cases. :see_no_evil: 
And it's not annoying at all. I would be very disappointed with myself if the code was merged and it does not work as desired, so I'm glad you found this problem.

Unfortunately I did not have any spare time since the last commit (that's why I haven't replied yet). But now I picked up the work again on this feature.

(And I think I understand now why nobody else tackled this seemingly small code change for this long requested feature :laughing: )

**Comment by KriSchu** [comment]
Hi everyone,

I just pushed my latest code changes. 
The check_disk plugin reports fs utilization now, even when one or multiple disks cannot be found and are ignored by option`--ignore-missing`.

I had to modify a bit in the part where the output gets generated including the preamble and I added a part in the output where ignored paths are listed, so that users understand, why the check does not print any information for the given `-p` parameter.

Here are some test results:
```
# When no path matches, there is a new message for that now.
[plugins]$ ./check_disk --ignore-missing -w 0% -c 0% -p /bar
DISK OK - No disks were found for provided parameters; - ignored paths: /bar;|

[plugins]$ ./check_disk --ignore-missing -w 0% -c 0% -p /home -p /bar
DISK OK - free space: /home 364632MiB (22% inode=99%); - ignored paths: /bar;| /home=1294399963136B;1762136162304;1762136162304;0;1762136162304

[plugins]$ ./check_disk --ignore-missing -w 0% -c 0% -p /home -p /bar -r /boot
DISK OK - free space: /home 364632MiB (22% inode=99%); /boot/efi 299MiB (99% inode=-); - ignored paths: /bar;| /home=1294399963136B;1762136162304;1762136162304;0;1762136162304 /boot/efi=0B;313524224;313524224;0;313524224


# Notice that only /foo is reported as ignored path and not /bar, because without -E option 
# they are treated as belonging to the same fs '/' and the second is not even added to the list 
# of fs to check (plugin behaved like that before already)
[plugins]$ ./check_disk --ignore-missing -w 0% -c 0% -p /foo -p /bar -r /boot
DISK OK - free space: /boot/efi 299MiB (99% inode=-); - ignored paths: /foo;| /boot/efi=0B;313524224;313524224;0;313524224

# With -E option all paths are reported as missing:
[plugins]$ ./check_disk --ignore-missing -w 0% -c 0% -E -p /foo -p /bar -r /boot
DISK OK - free space: /boot/efi 299MiB (99% inode=-); - ignored paths: /foo; /bar;| /boot/efi=0B;313524224;313524224;0;313524224

[plugins]$ ./check_disk --ignore-missing -w 0% -c 0% -E -p /foo -p /bar -p /boot
DISK OK - No disks were found for provided parameters; - ignored paths: /foo; /bar; /boot;|

# Unfortunately for -r option (regex-path), the ignored paths cannot be
# printed in the output. The code structure doesn't allow that right now.
[plugins]$ ./check_disk --ignore-missing -w 0% -c 0% -E -r '^/foo$' -r '^/bar$'
DISK OK - No disks were found for provided parameters;|

[plugins]$ ./check_disk --ignore-missing -w 0% -c 0% -E -p '/foo' -p / -p /home -i '/home'
DISK OK - free space: / 90775MiB (51% inode=94%); - ignored paths: /foo;| /=90221576192B;194964881408;194964881408;0;194964881408

[plugins]$ ./check_disk --ignore-missing -w 0% -c 0% --group="TEST" -r /foo -r '^/bar$' -p /foobar
DISK OK - No disks were found for provided parameters; - ignored paths: /foobar;|
```

Test with `-e` option (=report only errors):
```
[plugins]$ ./check_disk --ignore-missing -w 0% -c 0% -E -p /foo -p /bar -p /boot -e
DISK OK - ignored paths: /foo; /bar; /boot;|

[plugins]$ ./check_disk --ignore-missing -w 0% -c 0% -E -p /foo -p /bar -p /boot/efi -e
DISK OK - ignored paths: /foo; /bar;| /boot/efi=0B;313524224;313524224;0;313524224

[plugins]$ ./check_disk --ignore-missing -w 0% -c 0% -E -p /foo -p / -p /home -i /home -e
DISK OK - ignored paths: /foo;| /=90221576192B;194964881408;194964881408;0;194964881408
```

I haven't run tests with remote-fs or any stale checks. Maybe somebody has a test setup for that and can easily test?
Please also run your own test to help me verify, that the code is bug free.

I'm also happy about a close look at the code and a (quick) code quality check - my C skills are very rusty. Thanks in advance for the effort.

**Comment by KriSchu** [comment]
@sni @RincewindsHat What do you think about the changes? Could you start the test pipeline so that I see if something fails in the tests?
Of course I would also be happy about some tests by you :)

**Review by RincewindsHat — CHANGES_REQUESTED** [review]
Thank you very much, that must have been a lot of work. I have some comments though, if that's ok.

File: plugins/check_disk.c [hunk 1]
```diff
@@ -166,10 +169,12 @@ main (int argc, char **argv)
   int result = STATE_UNKNOWN;
   int disk_result = STATE_UNKNOWN;
   char *output;
+  char *ignored;
   char *details;
   char *perf;
   char *perf_ilabel;
   char *preamble;
+  char *ignored_preamble;
```
[the remark this answers is not in the document]
Comment (KriSchu) [reply plugins/check_disk.c:177]: That was another thing where I didn't understand why the other variables were initialized like that and I just went along with it (as I also didn't want to introduce new issues). But yes, I can change that. Makes sense to me.

**Comment by KriSchu** [comment]
> Thank you very much, that must have been a lot of work. I have some comments though, if that's ok.

This one pulled me very deep into the rabbit whole, yes :laughing: 
And sure, I appreciate your feedback! I made the changes accordingly and pushed them already. If you find something else, please let me know.

File: plugins/check_disk.c [hunk 2]
```diff
@@ -140,6 +141,7 @@ int verbose = 0;
 int erronly = FALSE;
 int display_mntp = FALSE;
 int exact_match = FALSE;
+int ignore_missing = FALSE;
```
[the remark this answers is not in the document]
Comment (RincewindsHat) [reply plugins/check_disk.c]: The reason for that is (afaik) that the `bool` type only was added to the C language with C99 (which, I would guess, was introduced 1999) and by that time a lot of people got used to use `int`s and some preprocessor things for that.
So we have a lot of people who "just learned it that way" and also using C99 things meant your code would not compile on older compilers.
So, you avoid using `bool` for some time and .. well, here we are xD

**Review by RincewindsHat — APPROVED** [review]
Pretty sure this does what it should do now.

**Comment by RincewindsHat** [comment]
Thank you so much for all the work! Hope to see you here again soon and I do you a beer :-)

**Pull request merged**